feat: Added fenix support to the firefox-android extension runner

rpl opened the pull request:
This pull request applies the following tweaks to make it easier to test a webextensions on a GeckoView-based browsers:

- Added Fenix and GeckoView to the auto-detected APK names (so that the user doesn't need to explicitly select the APK if Fenix is the only installed one)
- Exposed the `--adb-discovery-timeout` parameter (already supported internally but never exposed to as a cli option, it can be used to change the maximum time to wait for the remote debugging server discovery)
- Added a new `--firefox-apk-component` cli option (needed to make it easier to use `web-ext run` with other custom GeckoView-based apps, e.g. a user may add `--firefox-apk-component GeckoViewActivity` to run an extension on the geckoview_example)
- ~~Added a new --fennec-mode boolean parameter (which the user can set to "force disable" the "Fennec compatibility mode")~~ (removed as not strictly needed, we may add it in the future if it may actually help the transition from Fennec to Fenix)
- ~~Only require the READ_EXTERNAL_STORAGE permission for non-Fennec Firefox for Android browsers (required to be able to access the XPI file from the path where web-ext run has uploaded it on the device, the WRITE_EXTERNAL_STORAGE is only required on Fennec because it is executed on a separate Firefox profile created on the sdcard)~~ (geckoview_example doesn't seem to work without the WRITE_EXTERNAL_STORAGE, this PR keeps the requirement unmodified for now, we may re-evaluate later)

The version of the firefox-android extension runner in this pull request still defaults to use the "Fennec Compatibility mode" for APK files without ".fenix" in their name, the `--fennec-mode` CLI option is meant to allow the user to explicitly set the "Fennec Compatibility mode" if needed.
[Once Fenix would completely replace Fennec in the release channel](https://arewefenixyet.com/), the "Fennec Compatibility mode" should be changed to default to `false`.

[![web-ext run on Fenix - Screencast](https://img.youtube.com/vi/3V09P6CPsfY/0.jpg)](https://youtu.be/3V09P6CPsfY)

Follow ups on the GeckoView / Fenix side:
- [Bug 1596867 - [remote-dbg-next] Temporary extensions should be listed on remote devices](https://bugzilla.mozilla.org/show_bug.cgi?id=1596867)
  - to make it easier to debug the installed extensions (by making it visible from about:debugging for connected remote targets)
- [Bug 1614295 - Notify GeckoView embedder apps of new temporary extensions installed through the Remote Debugging Server](https://bugzilla.mozilla.org/show_bug.cgi?id=1614295)
  - otherwise the parts of the extensions that require a native UI (e.g. browserAction/pageAction, 
    options pages) wouldn't be available when the extension installation flow isn't started by the
    GeckoView app itself
- [ ] issue #NNN - Fennec Addon Manager view should list the extension temporarily installed through the Remote Debugging Server
  - otherwise the developer may not be sure if the extension has been successfully installed,
    and there wouldn't be an easy way to reach and test an extension options ui page.
  -  (TODO this depends from Bug 1614195, file on github after initial feedback on Bug 1614195)

rpl commented:
> Does this work with generic GeckoView applications?

yes, it should work on other GeckoView-based applications, if they provide a way to enable the "Remote Debugging via USB".

rpl commented on src/extension-runners/firefox-android.js:
```diff
@@ -269,6 +273,21 @@ export class FirefoxAndroidExtensionRunner {
+    if (firefoxApk && firefoxApk.includes('.fenix')) {
+      return false;
+    }
+
```
> Is it possible to detect the application version before assuming Fenix? I.e. if <=68.x, assume Fennec, otherwise Fenix.

I looked into it a bit, but I haven't found a reasonable way to do that yet, `pm list packages` doesn't seem to provide version details for the listed packages.

Personally I would prefer to keep this pull request as simple as possible, so that we may even release it as part of the next minor version (which I haven't published yet).

Nevertheless I'm definitely interested on improving the Fenix support and auto-detection in followups.

Rob--W replied:
Fenix is a moving target, whereas we know what Fennec looks like. In the long term, we should be able to assume Fenix by default, unless.

This is not a blocker for this PR, so you can file a follow-up to address this.

Rob--W commented on src/program.js:
```diff
@@ -621,6 +627,14 @@ Example: $0 --help run.
+      'fennec-mode': {
+        describe: (
+          'Control the APK as a Fennec-based browser (defaults to true, ' +
+          'set to false if the APK is a geckoview-based browser)'
```
nit: `GeckoView`.

Why is the default encoded in a comment instead of `default: true`?

rpl replied:
> nit: GeckoView.

:+1:

> Why is the default encoded in a comment instead of default: true?

Because the default `fennecMode` value is actually meant to be `undefined` for now, and in that case we still defaults to the "Fennec compatibility mode" unless the APK name contains ".fenix" in its name.

> Fenix is a moving target, whereas we know what Fennec looks like. In the long term, we should be able to assume Fenix by default, unless.
> ...

Once we are going to be ready to "assume Fenix by default", we can change change the `fennecMode` option definition to have `false` as its default value, and the user may only need to use it to force the legacy "Fennec compatibility mode" if ever necessary.

I'm going to file an issue, to make this change when we are ready to default to `fennecMode: false`.

rpl commented:
@Rob--W I added some additional changes to allow web-ext run to also work with the geckoview example (in particular I had to add an additional --firefox-apk-component option to allow a user to specify a different Android component to run).

I have also added back the write permission requirement (at least for now), without it geckoview doesn't seem to be starting correctly (I'm quite surprised about that issue, but I have to dig a bit more into it to find exactly what is going on).

rpl commented:
@Rob--W I've just updated this pull request as we agreed over Matrix (in particular by removing the --fennec-mode cli option, and update the test cases accordingly).
